src/Common/tests/Collections/IListTest.cs

```diff
@@ -0,0 +1,1692 @@
+                            list[i] = GenerateItem();
+                    }
+
+                    for (int i = 0; i < items.Length; i++)
+                    {
+                        list[i] = items[i];
+                    }
+
+                    CollectionAssert.Equal(items, list);
+                }
+            }
+        }
+
+        [Fact]
+        public void AddFixedSizeOrReadonlyThrows()
+        {
+            if (ExpectedIsFixedSize || ExpectedIsReadOnly)
+            {
+                var items = GenerateItems(10);
+                var list = GetList(items);
+                Assert.Throws<NotSupportedException>(
+                    () => list.Add(GenerateItem()));
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddWithNullValueInMiddle()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                object[] items;
+                var list = GetList(null);
+                var tempItems = GenerateItems(32);
+                for (int i = 0; i < tempItems.Length/2; i++)
+                {
+                    list.Add(tempItems[i]);
+                }
+
+                if (ItemsMustBeNonNull)
+                {
+                    items = tempItems;
+                    Assert.Throws<ArgumentNullException>(
+                        () => list.Add(null));
+                }
+                else
+                {
+                    items = new object[33];
+                    list.Add(null);
+                    items[16] = null;
+                    Array.Copy(tempItems, 0, items, 0, 16);
+                    Array.Copy(tempItems, 16, items, 17, 16);
+                }
+                for (int i = 0; i < 16; i++)
+                {
+                    list.Add(tempItems[16 + i]);
+                }
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddWithNullValueAtBeginning()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                var list = GetList(null);
+                object[] items;
+                if (ItemsMustBeNonNull)
+                {
+                    Assert.Throws<ArgumentNullException>(
+                        () => list.Add(null));
+                    items = GenerateItems(16);
+                    list.AddRange(items);
+                }
+                else
+                {
+                    items = new object[17];
+                    list.Add(null);
+                    items[0] = null;
+                    var tempItems = GenerateItems(16);
+                    list.AddRange(tempItems);
+                    Array.Copy(tempItems, 0, items, 1, 16);
+                }
+
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddWithNullValueAtEnd()
```

Comment:
The original test had cases for each, I think it has value for specialized collections, they might have special cases for nulls. It doesn't add much complexity either.